cabal-install-solver/src/Distribution/Solver/Types/ProjectConfigPath.hs

```diff
@@ -0,0 +1,216 @@
+--                └─ hops/hops-5.config
+--                   └─ ../hops-6.config
+--                      └─ hops/hops-7.config
+--                         └─ ../hops-8.config
+--                            └─ hops/hops-9.config
+-- @
+--
+-- Into paths like this:
+-- @
+-- └─ hops-0.project
+--    └─ hops/hops-1.config
+--       └─ hops-2.config
+--          └─ hops/hops-3.config
+--             └─ hops-4.config
+--                └─ hops/hops-5.config
+--                   └─ hops-6.config
+--                      └─ hops/hops-7.config
+--                         └─ hops-8.config
+--                            └─ hops/hops-9.config
+-- @
+--
+-- That way we have @hops-8.config@ instead of
+-- @./hops/../hops/../hops/../hops/../hops-8.config@.
+--
+-- Let's see how @canonicalizePath@ works that is used in the implementation
+-- then we'll see how @canonicalizeConfigPath@ works.
+--
+-- >>> let AbsoluteDir d = testDir
+-- >>> makeRelative d <$> canonicalizePath (d </> "hops/../hops/../hops/../hops/../hops-8.config")
+-- "hops-8.config"
+--
+-- >>> let ad@(AbsoluteDir d) = testDir
+-- >>> p <- canonicalizeConfigPath ad (ProjectConfigPath $ (d </> "hops/../hops/../hops/../hops/../hops-8.config") :| [])
+-- >>> render $ docProjectConfigPath p
+-- "hops-8.config"
+--
+-- >>> :{
+--   do
+--     let expected = unlines
+--           [ "hops/hops-9.config"
+--           , "  imported by: hops-8.config"
+--           , "  imported by: hops/hops-7.config"
+--           , "  imported by: hops-6.config"
+--           , "  imported by: hops/hops-5.config"
+--           , "  imported by: hops-4.config"
+--           , "  imported by: hops/hops-3.config"
+--           , "  imported by: hops-2.config"
+--           , "  imported by: hops/hops-1.config"
+--           , "  imported by: hops-0.project"
+--           ]
+--     let ad@(AbsoluteDir d) = testDir
+--     let configPath = ProjectConfigPath ("hops/hops-9.config" :|
+--           [ "../hops-8.config"
+--           , "hops/hops-7.config"
+--           , "../hops-6.config"
+--           , "hops/hops-5.config"
+--           , "../hops-4.config"
+--           , "hops/hops-3.config"
+--           , "../hops-2.config"
+--           , "hops/hops-1.config"
+--           , d </> "hops-0.project"])
+--     p <- canonicalizeConfigPath ad configPath
+--     return $ expected == render (docProjectConfigPath p) ++ "\n"
+-- :}
+-- True
+canonicalizeConfigPath :: AbsoluteDir -> ProjectConfigPath -> IO ProjectConfigPath
+canonicalizeConfigPath (AbsoluteDir d) (ProjectConfigPath p) = do
+    xs <- sequence $ NE.scanr (\importee -> (>>= \importer ->
+            if isURI importee
+                then pure importee
+                else canonicalizePath $ d </> takeDirectory importer </> importee))
+            (pure ".") p
+    return . makeRelativeConfigPath d . ProjectConfigPath . NE.fromList $ NE.init xs
+
+isURI :: FilePath -> Bool
+isURI = isJust  .parseURI
+
+newtype AbsoluteDir = AbsoluteDir FilePath
```

Comment:
This type could have a name that is more specific to its purpose, such as `AbsoluteProjectDir`, to distinguish it from a general purpose file path.